From fa3638a353040ce269bd9ad9f7f2d490083038df Mon Sep 17 00:00:00 2001
From: Harald Hoyer <harald@redhat.com>
Date: Wed, 27 May 2015 13:12:02 +0200
Subject: [PATCH] network: move ibft parsing to own script before main ip opt
 parsing

This way ibft can generate cmdline params for vlan to be parsed by
parse-vlan.
---
 modules.d/40network/install       |  1 +
 modules.d/40network/parse-bond.sh |  0
 modules.d/40network/parse-ibft.sh | 68 +++++++++++++++++++++++++++++++++++++++
 3 files changed, 69 insertions(+)
 mode change 100644 => 100755 modules.d/40network/parse-bond.sh
 create mode 100755 modules.d/40network/parse-ibft.sh

diff --git a/modules.d/40network/install b/modules.d/40network/install
index 348ef0b..be06f93 100755
--- a/modules.d/40network/install
+++ b/modules.d/40network/install
@@ -11,6 +11,7 @@ inst_hook pre-udev 50 "$moddir/ifname-genrules.sh"
 inst_hook pre-udev 60 "$moddir/net-genrules.sh"
 inst_hook cmdline 91 "$moddir/dhcp-root.sh"
 
+inst_hook cmdline 94 "$moddir/parse-ibft.sh"
 inst_hook cmdline 95 "$moddir/parse-vlan.sh"
 inst_hook cmdline 97 "$moddir/parse-bond.sh"
 inst_hook cmdline 98 "$moddir/parse-bridge.sh"
diff --git a/modules.d/40network/parse-bond.sh b/modules.d/40network/parse-bond.sh
old mode 100644
new mode 100755
diff --git a/modules.d/40network/parse-ibft.sh b/modules.d/40network/parse-ibft.sh
new file mode 100755
index 0000000..2c2cdae
--- /dev/null
+++ b/modules.d/40network/parse-ibft.sh
@@ -0,0 +1,68 @@
+#!/bin/sh
+# -*- mode: shell-script; indent-tabs-mode: nil; sh-basic-offset: 4; -*-
+# ex: ts=8 sw=4 sts=4 et filetype=sh
+
+. /lib/dracut-lib.sh
+
+for p in $(getargs ip=); do
+    [ "ibft" = "$p" ] || continue
+
+    modprobe iscsi_ibft
+    num=0
+    (
+        for iface in /sys/firmware/ibft/ethernet*; do
+            unset ifname_mac
+            unset ifname_if
+            unset dhcp
+            unset ip
+            unset gw
+            unset mask
+            unset hostname
+            unset vlan
+            [ -e ${iface}/mac ] || continue
+            ifname_mac=$(read a < ${iface}/mac; echo $a)
+            [ -z "$ifname_mac" ] && continue
+            unset dev
+            for ifname in $(getargs ifname=); do
+                if strstr "$ifname" "$ifname_mac"; then
+                    dev=${ifname%%:*}
+                    break
+                fi
+            done
+            if [ -z "$dev" ]; then
+                ifname_if=ibft$num
+                num=$(( $num + 1 ))
+                echo "ifname=$ifname_if:$ifname_mac"
+                dev=$ifname_if
+            fi
+
+            [ -e ${iface}/dhcp ] && dhcp=$(read a < ${iface}/dhcp; echo $a)
+            if [ -n "$dhcp" ]; then
+                echo "ip=$dev:dhcp"
+            else
+                [ -e ${iface}/ip-addr ] && ip=$(read a < ${iface}/ip-addr; echo $a)
+                [ "$ip" = "0.0.0.0" ] && unset ip
+                [ -e ${iface}/gateway ] && gw=$(read a < ${iface}/gateway; echo $a)
+                [ -e ${iface}/subnet-mask ] && mask=$(read a < ${iface}/subnet-mask; echo $a)
+                [ -e ${iface}/hostname ] && hostname=$(read a < ${iface}/hostname; echo $a)
+                [ -n "$ip" ] && echo "ip=$ip::$gw:$mask:$hostname:$dev:none"
+            fi
+
+            if [ -e ${iface}/vlan ]; then
+                vlan=$(read a < ${iface}/vlan; echo $a)
+                if [ "$vlan" -ne "0" ]; then
+                    case "$vlan" in
+                        [0-9]*)
+                            echo "vlan=$dev.$vlan:$dev"
+                            ;;
+                        *)
+                            echo "vlan=$vlan:$dev"
+                            ;;
+                    esac
+                fi
+            fi
+        done
+    ) >> /etc/cmdline
+    # reread cmdline
+    unset CMDLINE
+done
